types/src/bn254_circom.rs

```diff
@@ -2,42 +2,141 @@
 
 use crate::{
     jwks::rsa::RSA_JWK,
-    zkid::{
-        ZkIdPublicKey, ZkIdSignature, MAX_EPK_BYTES, MAX_EXPIRY_HORIZON_SECS, MAX_ISS_BYTES,
-        MAX_JWT_HEADER_BYTES,
-    },
+    zkid::{ZkIdPublicKey, ZkIdSignature, ZkpOrOpenIdSig, MAX_AUD_VAL_BYTES, MAX_EPK_BYTES, MAX_EXTRA_FIELD_BYTES, MAX_ISS_BYTES, MAX_JWT_HEADER_BYTES},
 };
 use anyhow::bail;
 use aptos_crypto::{poseidon_bn254, CryptoMaterialError};
-use ark_bn254::{Fq, Fq2};
+use ark_bn254::{Bn254, Fq, Fq2, Fr, G1Affine, G1Projective, G2Affine, G2Projective};
 use ark_ff::PrimeField;
 use ark_groth16::{PreparedVerifyingKey, VerifyingKey};
 use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use move_core_types::{ident_str, identifier::IdentStr, move_resource::MoveStructType};
 use once_cell::sync::Lazy;
 use serde::{Deserialize, Serialize};
 use serde_big_array::BigArray;
+use std::fmt::{Display, Formatter};
 
 // TODO(zkid): Some of this stuff, if not all, belongs to the aptos-crypto crate
 
 pub const G1_PROJECTIVE_COMPRESSED_NUM_BYTES: usize = 32;
 pub const G2_PROJECTIVE_COMPRESSED_NUM_BYTES: usize = 64;
 
-pub static DEVNET_VERIFYING_KEY: Lazy<PreparedVerifyingKey<ark_bn254::Bn254>> =
-    Lazy::new(devnet_pvk);
+/// Useful macro for arkworks serialization!
+macro_rules! serialize {
+    ($obj:expr) => {{
+        let mut buf = vec![];
+        $obj.serialize_compressed(&mut buf).unwrap();
+        buf
+    }};
+}
+
+/// Reflection of aptos_framework::zkid::Groth16PreparedVerificationKey
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Groth16VerificationKey {
+    /// Serialization of `alpha * G`, where `G` is the generator of `G1`.
+    alpha_g1: Vec<u8>,
+    /// Serialization of `alpha * H`, where `H` is the generator of `G2`.
+    beta_g2: Vec<u8>,
+    /// Serialization of `gamma * H`, where `H` is the generator of `G2`.
+    gamma_g2: Vec<u8>,
+    /// Serialization of `delta * H`, where `H` is the generator of `G2`.
+    delta_g2: Vec<u8>,
+    /// Serialization of `\forall i \in {0, 1}, gamma^{-1} * (beta * a_i + alpha * b_i + c_i) * H`, where `H` is the generator of `G1`.
```

Comment:
Should be forall i in {0, \ell} where \ell is the zkid circuit public input size, in this case 1. Spelling this out makes it easier to connect the code with the Groth16 math